src/libbson/src/bson/bson.c

```diff
@@ -415,6 +415,20 @@ _bson_append (bson_t *bson,              /* IN */
    return ok;
 }
 
+#define HANDLE_KEY_LENGTH(key, key_length)                                \
+   do {                                                                   \
+      if (key_length < 0) {                                               \
+         key_length = (int) strlen (key);                                 \
+      } else {                                                            \
+         /* Necessary to validate embedded NULL is not present in key. */ \
```

Comment:
Would it be simpler to return false if `key_length != strlen(key)`?

Reply:
I think that would assume that `key` is NULL-terminated, which it might not be.

---

Although I think moving this loop into a dedicated function, i.e.

```c
bool _string_contains_null(const char* str, size_t len) {
  for (; len; ++str, --len)
    if (*str == 0) return true;
  return false;
}
```

might be useful.

Reply:
The check has been simplified to `if (key_length != strlen (key))`. I remember you mentioned this to me before @jmikola but this slipped my mind 🤦.